lib/srv/db/common/databaseobjectimportrule/import_rules.go

```diff
@@ -0,0 +1,279 @@
+
+import (
+	"regexp"
+	"sort"
+	"strings"
+	"unicode"
+
+	"github.com/gravitational/trace"
+
+	dbobjectv1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobject/v1"
+	dbobjectimportrulev1 "github.com/gravitational/teleport/api/gen/proto/go/teleport/dbobjectimportrule/v1"
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/api/types/label"
+	"github.com/gravitational/teleport/api/utils"
+	"github.com/gravitational/teleport/lib/services"
+	libutils "github.com/gravitational/teleport/lib/utils"
+	"github.com/gravitational/teleport/lib/utils/typical"
+)
+
+// ApplyDatabaseObjectImportRules applies the given set of rules onto a set of objects coming from a same database.
+// Returns a fresh copy of a subset of supplied objects, filtered and modified.
+// For the object to be returned, it must match at least one rule.
+// The modification consists of application of extra labels, per matching mappings.
+// If there are any errors due to invalid label template, the corresponding objects will be dropped.
+func ApplyDatabaseObjectImportRules(rules []*dbobjectimportrulev1.DatabaseObjectImportRule, database types.Database, objs []*dbobjectv1.DatabaseObject) ([]*dbobjectv1.DatabaseObject, map[string]error) {
+	// sort: rules with higher priorities are applied last.
+	sort.Slice(rules, func(i, j int) bool {
+		return rules[i].Spec.Priority < rules[j].Spec.Priority
+	})
+
+	// filter rules: keep those with matching labels
+	// we only need mappings from the rules, so extract those.
+	var mappings []*dbobjectimportrulev1.DatabaseObjectImportRuleMapping
+	for _, rule := range rules {
+		dbLabels := make(types.Labels)
+		mapLabel := label.ToMap(rule.Spec.GetDatabaseLabels())
+		for k, v := range mapLabel {
+			dbLabels[k] = v
+		}
+		if ok, _, _ := services.MatchLabels(dbLabels, database.GetAllLabels()); ok {
+			mappings = append(mappings, rule.Spec.Mappings...)
+		}
+	}
+
+	var objects []*dbobjectv1.DatabaseObject
+	errors := map[string]error{}
+
+	// anything to do?
+	if len(mappings) == 0 {
+		return objects, errors
+	}
+
+	// find all objects that match any of the rules
+	for _, obj := range objs {
+		// prepare object clone
+		objClone := utils.CloneProtoMsg(obj)
+		if objClone.Metadata.Labels == nil {
+			objClone.Metadata.Labels = map[string]string{}
+		}
+
+		// apply each mapping in order.
+		matched := false
+		hadError := false
+		for _, mapping := range mappings {
+			match, err := applyMappingToObject(mapping, objClone.GetSpec(), objClone.Metadata.Labels)
+			if err != nil {
+				errors[obj.GetMetadata().GetName()] = trace.Wrap(err)
+				hadError = true
+				break
+			}
+			if match {
+				matched = true
+			}
+		}
+
+		if !hadError && matched {
+			objects = append(objects, objClone)
+		}
+	}
+
+	return objects, errors
+}
+
+// validateTemplate evaluates the template, checking for potential errors.
+func validateTemplate(template string) error {
+	_, err := evalTemplate(template, nil)
```

Comment:
the `nil` scared me for a second. so this works because the generated functions will return empty `""` when spec is `nil`? may worth leaving a comment or use an empty spec

Reply:
This works because `func (e expression) eval(spec *dbobjectv1.DatabaseObjectSpec) (string, error)` is resilient against null `spec`... which is a somewhat fragile property, now that I think about it. I'll pass an empty spec instead.